Universal plugins [wip]

RandomShaper opened the pull request:
## Goal

The main idea is creating a system that allows plugins to be installed into a specific project or globally (at the OS user level) so they are available for every project.

## Remarks

- Each commit adds a logical step on top of which others are built.
- Given the big scope of this PR, it's better to review it on a per-commit basis.
- Some of them (especially, the one marked with `[f]`) are meant to be squashed together before merging.
- Feedback and testing is welcome!

## Materials

### Specification

Please read the full [design document](https://docs.google.com/document/d/1fxHGnqPGCKFlxfjNw8T3W7r8BgQZlMQhWZNDH9LX3Hk/edit?usp=sharing).

The most recent commit (_Improve UX of universal plugins_) is not covered by the design document. It does a number of changes (detailed in the full commit message) to make the feature as easy-to-use as possible).

### Screenshots

- _Legacy Plugin_ is a legacy (`model=1`, the default) plugin at `<project_dir>/addons/`.
- _Universal Plugin A_ is a universal (`model=2`) plugin at `<project_dir>/addons/` (project-level, available to this project only).
- _Universal Plugin B_ is a universal (`model=2`) plugin at `<godot_config_dir>/addons/` (user-level, available to all projects).
- _Universal Plugin C_ is a universal (`model=2`) plugin packed as a PCK file at `<godot_config_dir>/addons/` (user-level, available to all projects).

## Plugin settings

![image](https://user-images.githubusercontent.com/11797174/72881827-cce45880-3d01-11ea-8e2a-27efd6b25661.png)

## Plugin creation dialog

![image](https://user-images.githubusercontent.com/11797174/72882607-2f8a2400-3d03-11ea-8bba-d3bc1f93b5c2.png)

## Default plugin code

![image](https://user-images.githubusercontent.com/11797174/72883025-dff82800-3d03-11ea-8751-9042573d187f.png)
(The explanation is added only for `model=2` plugins.)

## File system dock

![image](https://user-images.githubusercontent.com/11797174/72881951-0a48e600-3d02-11ea-9ab7-6e014c6d202b.png)

## File dialog

### Browsing resources

![image](https://user-images.githubusercontent.com/11797174/72882727-64967680-3d03-11ea-9de8-b23b796378dd.png)

## Directory dialog

![image](https://user-images.githubusercontent.com/11797174/72883250-45e4af80-3d04-11ea-938f-387300b7511e.png)
---
**This code is generously donated by IMVU.**

Commit:
Add ConfigFile::parse()

Commit:
Improve UX of drive letters

Namely, move the drive dropdown to just the left of the path text box and don't include the former
in the latter.

This by itself improves the UX on Windows, but it sets the ground for the following changes.

In the UNIX case, since its concept of drives is (ab)used to provide shortcuts to useful paths, its
dropdown is kept at the original location.

Commit:
Encapsulate import settings path generation

Commit:
Make file system scan more robust

Previously, a change was missed if it happened while the scan was in progress and already past the affected location.

Also:
- Consider the scan changes thread on termination, in addition to the full scan one
- Add FS-reported hidden to the check for hidden by the editor file system

Commit:
Ignore hidden files and directoreis in find in files

Commit:
Centralize resource access in the new `(Dir/File)AccessResources`

First, removed the notion of logical paths from the FS-specific classes so that they deal only with
physical paths. Therefore, translating back and forth between the abstracted (like `res://...`) and
the real one will happen in the base classes. They were already doing part of that work, but the
split wasn't clear enough.

On top of that, `(Dir/File)AccessResources` are introduced to be the ones specifically handling
access to resources. `DirAccessResources` has the notion of drives, being `res://` the only one for
the time being.

`String::simplify_path()` is made simpler and more universal.

`String::is_abs_path()` now also considers paths with a leading Windows drive.

Commit:
Generalize resource access

Make the full engine and editor aware of the fact that resources are not necessarily located
under `res://`.

Commit:
Fix VariantParser::StreamString EOF determination

Commit:
Improve/fix packed data API

- Add support for multiple `PackedData` instances
- Enhance directory API
- Fix `FileAccess::exists()` not checking for PackedData being disabled
- Fix `DirAccessPack::change_dir()` to handle ".." properyly (before `simplify_path()` disallows it)
- Allow absolute paths in existence checks

Commit:
Implement pack unmounting

Commit:
Add SIMPLE_PACK export mode

Commit:
Implement the new plugin model

Highlights:
- `addons://` is introduced as a virtual resources drive.
- Plugin creation dialog:
  - New model is the default option.
  - If created with new model, the default script contains an explanation about it.
  - Placeholders containing paths are adjusted to the model currently selected.
- If on rescan an enabled plugin is found not to be the same (for instance, a plugin is copied into
  the project while a user-wide one of the same name is enabled), the old one is teared down and
  the new is started up, making the switch seamless.
- If a legacy plugin is located in the editor settings, since that won't work, it's ignored and a
  warning about it is printed.

Commit:
Fully abstract `DirAccessResources` [f]

Makes it able to seamlessly jump between packed and non-packed locations.
Therefore, this commits also reverts what the editor file system scan needed before
this was done.

Commit:
Improve UX of universal plugins

FS dock:
- Disallow impossible drag & drop cases
- Handle _Show in file manager_ gracefully for items in PCK
- Handle move, rename and remove gracefully for items in PCK
- Hide the _New folder/resource/script_ and _Edit dependencies_ context menu items for read-only folders (PCK or `addons://`)
- Add _[PCK]_ to every PCK plugin's root name
- Paint items in PCK as grayed out
- Add an informative tooltip about the location for children of `addons://`

`EditorDirDialog`:
- Paint items in PCK as grayed out
- Add _[PCK]_ to every PCK plugin's root
- Disable the _Create folder_ button for read-only folders (PCK or `addons://`)
- Disallow picking a read-only folder as a destination path

`EditorFileDialog`:
- Disallow create folder and delete items if read-only (PCK or `addons://`)
- Disallow picking a read-only folder in save mode
- Paint items in PCK as grayed out

Other:
- Force correct PCK filename when exporting a plugin

YeldhamDev commented on editor/plugin_config_dialog.cpp:
```diff
@@ -78,17 +111,30 @@ void PluginConfigDialog::_on_confirmed() {
+						TTR("#   In either case it will be available in the Godot file system as addons://plugin_name.\n") +
+						TTR("# - To reference resources, you must use that kind of paths. For instance:\n") +
+						"#   var scene = load(\"addons://my_plugin/scene.tscn\")\n"
+						"\n";
```
The comment should respect the 80 characters limit. Also, all other code comments created by the editor aren't translated, so with this the other should probably follow suit then.

HeadClot commented:
When this gets committed to the main branch does this mean that we will need to update our plugins to account for this change to plugins? Just curious :)

vnen commented:
> When this gets committed to the main branch does this mean that we will need to update our plugins to account for this change to plugins? Just curious :)

According to the linked design document, old plugins still work the same way.

RandomShaper commented:
> According to the linked design document, old plugins still work the same way.

That's right. Anyone not ready or willing to use the new model will enjoy complete backwards compatibility.

groud commented:
I think we should be able to display the scope of a plugin directly in the filesystem dock. That would avoid messing up with user-wide plugins in some case.

fire commented:
There's CICD failures here https://travis-ci.org/godotengine/godot/builds/640402635

RandomShaper commented:
> I think we should be able to display the scope of a plugin directly in the filesystem dock. That would avoid messing up with user-wide plugins in some case.

Good idea. Right now the tooltip you get for children of `addons://` gives information about where it is physically.

However, using a different color or folder icon would be nice. The plugin settings dialog is using the accent color for user-level. That "language" could be used there, too.

Commit:
Allow `Editor`/`FileDialog` to force a file name

If the file name text field is made not-editable, it won't be tried to be changed, keeping any text that has been set to it externally.

RandomShaper commented:
Not much of general interest. Closing.

arkology commented:
It'll be completely offtopic, but this is how Godot loses great features.

Xrayez commented:
@RandomShaper I hope a particular implementation will be discussed at [GIP](https://github.com/godotengine/godot-proposals) at least.

I also wonder what's the real reason for closing the PR, was there any discussion regarding this particular implementation? Not all users follow IRC discussions. There's certainly a general interest in this.

akien-mga commented:
I'm not sure why @RandomShaper's closing comment was so laconic, but to clarify, we discussed this at length at the past Godot Sprint in October 2019, and then again during GodotCon in February 2020. This is a wanted feature but no in its current form, which is needlessly complex.

A new proposal is expected building on this one, which keeps things simple while allowing having global editor plugins.
